### PR TITLE
refactor(transaction): Implement canonical ordering in `TransactionPlan`

### PR DESCRIPTION
**Pull request by TalDerei**
References #3435 and https://github.com/penumbra-zone/penumbra/issues/3379

### CODE REVIEW

**Comment by cronokirby** [comment]
We should also have tie breaking based on the effect hash, rather than just using whatever order between actions that happens to exist (or perhaps something more intricate). Because of this, it might be a good idea to instead shift the approach of this PR to adding an Eq and Ord instance to Action based on this philosophy.

**Comment by TalDerei** [comment]
> effect hash

Why is tie breaking against the effect hash necessary for the same action type?

**Comment by TalDerei** [comment]
This PR will remain as a draft until after the next release.

**Comment by conorsch** [comment]
We are not planning to merge this for Testnet 64, so we'll return to it next week.

**Comment by TalDerei** [comment]
Is this something we want done for the next testnet release? otherwise it can remain in the backlog

**Comment by hdevalence** [comment]
Let's keep it in the backlog for now

**Comment by TalDerei** [comment]
reopened since #3435 was tagged with `V1`

**Comment by TalDerei** [comment]
blocked until https://github.com/penumbra-zone/penumbra/issues/4081

**Comment by TalDerei** [comment]
unblocked by https://github.com/penumbra-zone/penumbra/pull/4312

File: crates/core/transaction/src/plan.rs [hunk 1]
```diff
@@ -48,6 +49,20 @@ pub struct TransactionPlan {
...
+
+        actions
+    }
+
+    /// Sort the action plans by type, using the protobuf field number in the [`ActionPlan`].
+    pub fn sort_action_plans(&self, mut actions_plans: Vec<ActionPlan>) -> Vec<ActionPlan> {
+        actions_plans.sort_by_key(|action: &ActionPlan| action.variant_index());
+
+        actions_plans
+    }
```
Comment (hdevalence) [inline crates/core/transaction/src/plan.rs:64]: I think we probably just want to have a single method
```
TransactionPlan::sort(&mut self) { ... }
```
right? Then we can sort the plan exactly once, immediately before the planner returns it, and then we won't have to change any of the other code.

Comment (TalDerei) [reply crates/core/transaction/src/plan.rs]: yeah that's much cleaner

**Review by hdevalence — APPROVED** [review]
LGTM!

**Pull request merged**